src/graph.rs

```diff
@@ -2134,10 +2329,149 @@ impl<'a, 'graph> Builder<'a, 'graph> {
       };
       if let (Some(specifier), Some(reporter)) = (specifier, self.reporter) {
         let modules_total = self.graph.module_slots.len();
-        let modules_done = modules_total - self.pending.len();
+        let modules_done = modules_total - self.state.pending.len();
         reporter.on_load(&specifier, modules_done, modules_total);
       }
-      if self.pending.is_empty() {
+
+      if self.state.pending.is_empty() {
+        // resolve all the pending deno specifiers
+        let pending_resolutions =
+          std::mem::take(&mut self.state.deno.pending_resolutions);
+        let mut pending_version_resolutions =
+          Vec::with_capacity(pending_resolutions.len());
+        for pending_resolution in pending_resolutions {
+          let package_name = &pending_resolution.package_ref.req().name;
+          let fut = self
+            .state
+            .deno
+            .pending_package_info_loads
+            .get_mut(package_name)
+            .unwrap()
+            .clone();
+          match fut.await {
+            Ok(Some(info)) => {
+              // resolve the best version out of the existing versions first
+              let package_req = pending_resolution.package_ref.req();
+              match self.resolve_deno_version(&info, package_req) {
+                Some(version) => {
+                  // now queue a pending load for that version information
+                  let package_nv = PackageNv {
+                    name: package_name.clone(),
+                    version,
+                  };
+                  self
+                    .graph
+                    .deno_specifiers
+                    .add(package_req.clone(), package_nv.clone());
+
+                  self.queue_load_package_version_info(&package_nv);
+                  pending_version_resolutions
+                    .push((package_nv, pending_resolution));
+                }
+                None => {
+                  if self.fill_pass_mode == FillPassMode::AllowRestart {
+                    self.restart(provided_roots, provided_imports).await;
+                    return;
```

Comment:
Is there a limit how many restarts can happen? Is it possible it will get into an infinite restart loop?

Reply:
In `self.restart` it will set the fill_pass_mode to `FillPassMode::CacheBusting`, at which point it won't ever be `AllowRestart` again.